[lldb] Factor out common Swift Async Context utilities

kastiglione opened the pull request:
Extract functions related to swift concurrency for reuse.

1. Add `SwiftLanguageRuntime::GetAsyncContext()` which reads the async context address from the target specific register.
2. Add `GetAsyncUnwindRegisters()` which returns an optional struct that holds the target specific registers used for unwinding.

kastiglione committed:
[lldb] Factor out common Swift Async Context utilities

kastiglione commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2285,34 +2306,19 @@ GetFollowAsyncContextUnwindPlan(RegisterContext *regctx, ArchSpec &arch,
-    row->GetCFAValue().SetIsRegisterDereferenced(async_context_regnum);
-    row->SetRegisterLocationToSame(async_context_regnum, false);
+    row->GetCFAValue().SetIsRegisterDereferenced(registers->async_ctx_regnum);
+    row->SetRegisterLocationToSame(registers->async_ctx_regnum, false);
```
most of the change in this PR is noise, where regular local variables are now fields on structs. I could unpack the struct and return these all to using local vars, if preferred.

kastiglione commented:
@swift-ci test

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,50 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
 
 bool SwiftLanguageRuntime::IsABIStable() { FORWARD(IsABIStable); }
 
+struct AsyncUnwindRegisters {
```
Should we add some doxygen comments? Particularly dummy_reg doesn't seem to explain itself :-)

kastiglione replied:
maybe it can be named better, what does everyone think about `indirect_marker_regnum`? cc @fredriss

kastiglione replied:
(I added doc comments too)

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,50 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
+    return (AsyncUnwindRegisters){
+        .async_ctx_regnum = dwarf_r14_x86_64,
+        .fp_regnum = dwarf_rbp_x86_64,
+        .pc_regnum = dwarf_rip_x86_64,
```
Why do we need to store pc_regnum here, doesn't LLDB already know this elsewhere?

jasonmolenda replied:
we have LLDB_GENERIC register numbers, but they use a different register number space.  The UnwindPlans need to be expressed in a single consistent register numbering, so we need the eh_frame / dwarf register numbers for pc/fp.  Normally you'd be right, we could use the generic_fp and generic_pc.

jasonmolenda replied:
(there is no LLDB_GENERIC that would get you the async context pointer.  You could argue for adding one, tbh, that's another way to approach this.)

kastiglione replied:
I captured some of this conversation in the doc comments.

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,50 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
+        .async_ctx_regnum = dwarf_r14_x86_64,
+        .fp_regnum = dwarf_rbp_x86_64,
+        .pc_regnum = dwarf_rip_x86_64,
+        .dummy_regnum = dwarf_r15_x86_64,
```
It would be nice if we could get this definition from header in LLVM, but I guess there probably is none that would define the DWARF regnum for this, right? But it would be nice if this were defined in some tablegen-erated header in LLVM instead.

jasonmolenda replied:
to be fair, we use these constants in a half dozen places in lldb already.

adrian-prantl reviewed (commented):
This is definitely going in the right direction!

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,50 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
+        .dummy_regnum = arm64_dwarf::x23,
+    };
+  default:
+    assert(false && "swift async supports only x86_64 and arm64");
```
I think this may actually a proper use-case for lldb_assert()?

kastiglione replied:
That would be a non-NFC change. Currently `GetRuntimeUnwindPlan()` returns a null plan, which I think is better than crashing. It'd be better to show an OS backtrace instead of crashing.

jasonmolenda commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2285,34 +2306,19 @@ GetFollowAsyncContextUnwindPlan(RegisterContext *regctx, ArchSpec &arch,
-    pc_regnum = arm64_dwarf::pc;
-    dummy_regnum = arm64_dwarf::x23;
-  } else {
+  auto registers = GetAsyncUnwindRegisters(arch.GetMachine());
```
This is 100% a personal choice, but I think it makes it harder to read the code years from now with this kind of `auto` use.  Is this getting a register value?  Is it getting a register context?  What can I do with the returned thing, is it a pointer that can return nullptr?  Sometimes auto is a billion times better like for (auto it = vec.begin(), but sometimes I think it adds a burden to future maintainers who are reading it. If this were a common idiom that we saw, like auto target_sp(GetTarget()), instead of TargetSP, everyone would know what this was doing.

I'm not saying you should change this, just throwing out my personal opinion and I think I am not in the majority within the group, so feel free to leave it as-is.

kastiglione replied:
changed it, thanks

jasonmolenda commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,50 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
+};
+
+static llvm::Optional<AsyncUnwindRegisters>
+GetAsyncUnwindRegisters(llvm::Triple::ArchType triple) {
```
It might be clearer if it says GetAsyncUnwindRegisterNumbers.  But I do love overly long variable & function names.

kastiglione replied:
that's better, I changed it.

kastiglione committed:
1. add some headerdocs
2. use RegisterNumbers/regnums for the struct and its vars
3. declare with explicity type for readability

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,25 +2095,31 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
 
-struct AsyncUnwindRegisters {
+// The target specific register numbers (eh_frame/dwarf) used for async
+// unwinding.
```
Annoying nitpicking: using `///` here will format more nicely in some IDEs, even if it won't make it into the doxygen html because it's local.

kastiglione replied:
done

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,25 +2095,31 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
 
 bool SwiftLanguageRuntime::IsABIStable() { FORWARD(IsABIStable); }
 
-struct AsyncUnwindRegisters {
```
Should this be in an anonymous namespace?

kastiglione replied:
done

kastiglione committed:
1. use /// comments
2. put `AsyncUnwindRegisters` in an anon namespace

kastiglione committed:
clang-format

kastiglione commented:
@swift-ci test

kastiglione commented:
@swift-ci test macOS platform

drodriguez commented on lldb/source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -2095,6 +2095,59 @@ void SwiftLanguageRuntime::DidFinishExecutingUserExpression(
+GetAsyncUnwindRegisterNumbers(llvm::Triple::ArchType triple) {
+  switch (triple) {
+  case llvm::Triple::x86_64:
+    return (AsyncUnwindRegisterNumbers){
```
I think this syntax is C syntax that Clang might accept in C++, but it is not accepted by MSVC. Both VS2017 and VS2019 are failing in CI:

```
C:\jenkins\workspace\oss-swift-windows-x86_64\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2119): error C2059: syntax error: '.'
C:\jenkins\workspace\oss-swift-windows-x86_64\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2124): error C2046: illegal case
C:\jenkins\workspace\oss-swift-windows-x86_64\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2126): error C2059: syntax error: '.'
```

and

```
S:\jenkins\workspace\oss-swift-windows-x86_64-vs2019\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2119): error C7555: use of designated initializers requires at least '/std:c++latest'
S:\jenkins\workspace\oss-swift-windows-x86_64-vs2019\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2123): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax
S:\jenkins\workspace\oss-swift-windows-x86_64-vs2019\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2126): error C7555: use of designated initializers requires at least '/std:c++latest'
S:\jenkins\workspace\oss-swift-windows-x86_64-vs2019\llvm-project\lldb\source\Target\SwiftLanguageRuntime.cpp(2130): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax
```

I understand that this is more readable, but would anyone mind changing it to be standard C++? (I might had time to do it at some moment, but anyone feel free to change it before me)

kastiglione replied:
I'll check it thanks.

kastiglione replied:
https://github.com/apple/llvm-project/pull/2807

etcwilde replied:
Designated initializers are a part of C++20, so it's already part of the standard. Just not the standard we're using. :)
